servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,675 @@
+                        }
+                    } else if (afterLockSubscription == PROCESSING_REQUEST &&
+                            subscriptionUpdater.compareAndSet(this, PROCESSING_REQUEST, prevSubscription)) {
+                        break;
+                    }
+                }
+            }
+
+            private void cancelAndGiveBack(Subscription subscription) {
+                try {
+                    subscription.cancel();
+                } finally {
+                    giveBackUnusedRequestN();
+                }
+            }
+
+            private boolean giveBackUnusedRequestN() {
+                // we need to give back the outstanding amount that has been requested, but not emitted.
+                // resetting sourceRequestedUpdater must happen unconditionally first, as this stops
+                // calculateSourceRequested from attempting to request any more.
+                final long prevOutstandingDemand = outstandingDemandUpdater.getAndSet(this, -1);
+                return prevOutstandingDemand >= 0 && parent.removeSubscriber(this, prevOutstandingDemand);
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                for (;;) {
+                    final Subscription prevSubscription = subscription;
+                    assert prevSubscription != CANCEL_PENDING && prevSubscription != PROCESSING_ONNEXT &&
+                            prevSubscription != PROCESSING_REQUEST;
+                    if (prevSubscription == null) {
+                        if (subscriptionUpdater.compareAndSet(this, null, s)) {
+                            break;
+                        }
+                    } else if (prevSubscription == REQUEST_PENDING) {
+                        if (subscriptionUpdater.compareAndSet(this, REQUEST_PENDING, PROCESSING_REQUEST)) {
+                            doRequestMore(s);
+                            break;
+                        }
+                    } else {
+                        s.cancel(); // already cancelled or duplicate onSubscribe
+                        break;
+                    }
+                }
+            }
+
+            @Override
+            public void onNext(@Nullable final R r) {
+                boolean acquiredLock = false;
+                for (;;) {
+                    Subscription prevSubscription = subscription;
+                    assert prevSubscription != null;
+                    if (prevSubscription == CANCELLED || prevSubscription == CANCEL_PENDING) {
+                        // we have already given our undelivered requestN quota up, or will after we unroll process
+                        // onNext. we are not allowed to deliver more data or else we may violate upstream's requestN.
+                        break;
+                    } else if (prevSubscription == PROCESSING_ONNEXT ||
+                            (acquiredLock = subscriptionUpdater.compareAndSet(this,
```

Comment:
a key part to understanding the approach is as follows:
there is no queuing of signals for the purpose of managing excess demand [1]. instead the demand from downstream is carefully distributed to all mapped publishers. the "careful" part means that we need to be able to reclaim demand from mapped publishers when they terminated or are cancelled, otherwise we may deadlock due to not delivering data that was requested. bcz we need to reclaim demand upon cancellation we need to be sure there is no data delivered after this point (to avoid delivering too much data). So the `subscription` lock provides that coordination and makes sure we only deliver what we are allowed, and give it all back when we terminate/cancel.

[1] there is an internal queue to manage concurrency requirements of

Reply:
consider adding this as a comment in the code-base.

Reply:
agreed. I added it to the class level comments.